### PR TITLE
1118 manage profile on admin panel & psss

### PR DESCRIPTION
**Pull request by tcaiger**
### Changes:

- Add profile pages and routes
- Add profile form
- Add react hook form and form validation to form
- Add profile form handling
- tweak server to return more user fields

---

### Screenshots:
Screen Shot 2020-09-21 at 4 10 56 PM

### CODE REVIEW

File: packages/admin-panel/src/App.js [hunk 1]
```diff
@@ -8,7 +8,9 @@ import { BrowserRouter as Router, Route, Switch, Redirect } from 'react-router-d
 import { TabsToolbar } from '@tupaia/ui-components';
 import { Navbar, Footer } from './widgets';
 import { ROUTES } from './routes';
-import { PrivateRoute, LoginView } from './authentication';
+import { PROFILE_ROUTES } from './profileRoutes';
+import { PrivateRoute } from './authentication';
+import { LoginPage } from './pages/LoginPage';
```
Comment (edmofro) [inline packages/admin-panel/src/App.js:13]:
```suggestion
import { PrivateRoute } from './authentication';
import { PROFILE_ROUTES } from './profileRoutes';
import { LoginPage } from './pages/LoginPage';
```

File: packages/admin-panel/src/authentication/actions.js [hunk 2]
```diff
@@ -63,3 +64,17 @@ export const loginError = errorMessage => ({
 export const logout = () => ({
   type: LOGOUT,
 });
+
+// Profile
+export const updateProfile = (id, payload) => async (dispatch, getState, { api }) => {
+  try {
+    await api.put(`user/${id}`, null, payload);
```
Comment (edmofro) [inline packages/admin-panel/src/authentication/actions.js:71]: We should be able to use a `/me` endpoint here, which can then have built in permission checking that the same user is making the request as is being edited

File: packages/admin-panel/src/pages/ChangePasswordPage.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
+/**
+ * Tupaia MediTrak
```
Comment (edmofro) [inline packages/admin-panel/src/pages/ChangePasswordPage.js:2]: This file's not really part of this PR, but will hand wave it coz I think that's the next PR in the reviews channel

File: packages/meditrak-server/src/routes/authenticate.js [hunk 4]
```diff
@@ -57,6 +57,10 @@ const getAuthorizationObject = async ({ accessPolicy, refreshToken, user, permis
   const userDetails = {
     id: user.id,
     name: user.fullName,
+    firstName: user.first_name,
+    lastName: user.last_name,
+    position: user.position,
+    employer: user.employer,
```
Comment (edmofro) [inline packages/meditrak-server/src/routes/authenticate.js:63]: This adds the extra fields to every client doing authentication. I think I'd prefer to add a `GET /me` endpoint that just returns more info about the user attached to the auth header

File: packages/admin-panel/src/pages/ProfilePage.js [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+const ProfilePageComponent = React.memo(({ user, onUpdateProfile, getHeaderEl }) => {
+  const [errorMessage, setErrorMessage] = useState(null);
+  const [successMessage, setSuccessMessage] = useState(null);
+  const [isLoading, setIsLoading] = useState(false);
+  const { handleSubmit, register, errors } = useForm();
+  const HeaderPortal = usePortalWithCallback(<Header title={user.name} />, getHeaderEl);
+
+  const onSubmit = handleSubmit(async ({ firstName, lastName, role, employer }) => {
+    setIsLoading(true);
+    setErrorMessage(null);
```
Comment (edmofro) [inline packages/admin-panel/src/pages/ProfilePage.js:46]:
```suggestion
    setErrorMessage(null);
    setSuccessMessage(null);
```

File: packages/admin-panel/src/pages/ProfilePage.js [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        first_name: firstName,
+        last_name: lastName,
+        position: role,
+        employer: employer,
+      });
+      setIsLoading(false);
+      setSuccessMessage('Profile successfully updated.');
+    } catch (error) {
+      setErrorMessage(error.message);
+      setIsLoading(false);
```
Comment (edmofro) [inline packages/admin-panel/src/pages/ProfilePage.js:58]: This is super nitpicky, but just for consistency...

```suggestion
      setIsLoading(false);
      setSuccessMessage('Profile successfully updated.');
    } catch (error) {
      setIsLoading(false);
      setErrorMessage(error.message);
```

File: packages/admin-panel/src/pages/ProfilePage.js [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+const SuccessMessage = styled.p`
+  color: ${props => props.theme.palette.success.main};
+`;
+
+const ProfilePageComponent = React.memo(({ user, onUpdateProfile, getHeaderEl }) => {
+  const [errorMessage, setErrorMessage] = useState(null);
+  const [successMessage, setSuccessMessage] = useState(null);
+  const [isLoading, setIsLoading] = useState(false);
+  const { handleSubmit, register, errors } = useForm();
```
Comment (edmofro) [inline packages/admin-panel/src/pages/ProfilePage.js:41]: With `noValidate` on the form, will `errors` ever have content?

Comment (edmofro) [reply packages/admin-panel/src/pages/ProfilePage.js:41]: Oh yep, presumably the "required" fields will still have that validated

**Review by edmofro — CHANGES_REQUESTED** [review]
Front end all looks awesome. Just got some api change suggestions

File: packages/admin-panel/src/pages/ProfilePage.js (hunk 5 above)
Comment (tcaiger) [reply packages/admin-panel/src/pages/ProfilePage.js:46]: Good catch

File: packages/admin-panel/src/pages/ProfilePage.js (hunk 7 above)
Comment (tcaiger) [reply packages/admin-panel/src/pages/ProfilePage.js:41]: yes that's correct, it's just doing required validation in this instance

File: packages/meditrak-server/src/routes/authenticate.js (hunk 4 above)
Comment (tcaiger) [reply packages/meditrak-server/src/routes/authenticate.js:63]: hmmm I feel like there is no real downside to including a few extra fields on each response and it saves doing an extra request every time the user logs in. Plus if we keep using `name` in this endpoint and then have another endpoint with `firstName` and `lastName`, we would have to be careful not to let them get out of sync in an app. eg. user details have been changed in AdminPanel and PSSS user has not logged back out but has been re-authenticated in the background.

File: packages/meditrak-server/src/routes/editUser.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
+/**
+ * Tupaia MediTrak
+ * Copyright (c) 2017 Beyond Essential Systems Pty Ltd
+ */
```
Comment (edmofro) [inline packages/meditrak-server/src/routes/editUser.js:4]: Even inside meditrak-server we now use the generic mono-repo header for new files

(I only notice because of your updates to the others)

File: packages/meditrak-server/src/routes/editUser.js [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+  if (!user) {
+    throw new DatabaseError('User not found');
+  }
+
+  try {
+    req.params = {
+      id: userId,
+      resource: 'user',
+    };
+    await editRecord(req, res);
```
Comment (edmofro) [inline packages/meditrak-server/src/routes/editUser.js:21]: This will run into conflicts once @EMcQ-BES has finished the permissions work

File: packages/meditrak-server/src/routes/authenticate.js (hunk 4 above)
Comment (edmofro) [reply packages/meditrak-server/src/routes/authenticate.js:63]: Fair enough, I'm happy with that :-)

**Review by edmofro — APPROVED** [review]
Approved, just a copyright header to change over before merging
